rosbag2_transport/test/rosbag2_transport/test_record.cpp

```diff
@@ -126,10 +137,21 @@ TEST_F(RecordIntegrationTestFixture, can_record_again_after_stop)
     std::move(writer_), storage_options_, record_options);
   recorder->record();
 
+  start_async_spin(recorder);
+
+  // Wait until recorder discovery is complete, otherwise messages might be missed.
+  // The currently expected topics:
+  // /string_topic
+  auto discovery_ret = rosbag2_test_common::wait_until_condition(
+    [&recorder]() {
+      return recorder->subscriptions().size() == 1;
+    },
+    std::chrono::seconds(5));
+  ASSERT_TRUE(discovery_ret);
+
   auto & writer = recorder->get_writer_handle();
   auto & mock_writer = dynamic_cast<MockSequentialWriter &>(writer.get_implementation_handle());
 
-  start_async_spin(recorder);
   ASSERT_TRUE(pub_manager.wait_for_matched(string_topic.c_str()));
```

Comment:
Do we really need to call `wait_until_condition`? here it waits for the one subscription matched to this publisher, that means publisher confirms the corresponding subscriber to send the message. (this goes to some other changes in this PR.)

Reply:
And now fixed in 02d5bba